scalapact-core/src/main/scala/com/itv/scalapactcore/common/matching/InteractionMatchingPrograms.scala

```diff
@@ -0,0 +1,213 @@
+case object MatchStatusFailed extends MatchStatusOutcome
+
+sealed trait MatchHeadersOutcome
+case object MatchHeadersSuccess extends MatchHeadersOutcome
+case object MatchHeadersFailed extends MatchHeadersOutcome //TODO: What kind of match failures do we get that we can model?
+
+sealed trait MatchBodyOutcome
+case object MatchBodySuccess extends MatchBodyOutcome
+case object MatchBodyFailed extends MatchBodyOutcome //TODO: What kind of match failures do we get that we can model?
+
+sealed trait MatchMethodOutcome
+case object MatchMethodSuccess extends MatchMethodOutcome
+case object MatchMethodFailed extends MatchMethodOutcome //TODO: What kind of match failures do we get that we can model?
+
+sealed trait MatchPathOutcome
+case object MatchPathSuccess extends MatchPathOutcome
+case object MatchPathFailed extends MatchPathOutcome //TODO: What kind of match failures do we get that we can model?
+
+sealed trait MatchResponseStageA[A]
+
+case class MatchResponseStatus(expected: InteractionResponse, actual: InteractionResponse) extends MatchResponseStageA[MatchStatusOutcome]
+case class MatchResponseHeaders(expected: InteractionResponse, actual: InteractionResponse) extends MatchResponseStageA[MatchHeadersOutcome]
+case class MatchResponseBody(expected: InteractionResponse, actual: InteractionResponse) extends MatchResponseStageA[MatchBodyOutcome]
+
+
+sealed trait MatchRequestStageA[A]
+
+case class MatchRequestMethod(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchMethodOutcome]
+case class MatchRequestPath(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchPathOutcome]
+case class MatchRequestHeaders(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchHeadersOutcome]
+case class MatchRequestBody(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchBodyOutcome]
+
+case class RequestMatchOutcome(method: MatchMethodOutcome, path: MatchPathOutcome, headers: MatchHeadersOutcome, body: MatchBodyOutcome) {
+  val success: Boolean = {
+    (method, path, headers, body) match {
+      case (MatchMethodSuccess, MatchPathSuccess, MatchHeadersSuccess, MatchBodySuccess) => true
+      case _ => false
+    }
+  }
+
+  val failed: Boolean = !success
+}
+
+case class ResponseMatchOutcome(status: MatchStatusOutcome, headers: MatchHeadersOutcome, body: MatchBodyOutcome) {
+  val success: Boolean = {
+    (status, headers, body) match {
+      case (MatchStatusSuccess, MatchHeadersSuccess, MatchBodySuccess) => true
+      case _ => false
+    }
+  }
+
+  val failed: Boolean = !success
+}
+
+object InteractionMatchingPrograms {
+
+  type MatchResponseStage[A] = Free[MatchResponseStageA, A]
+
+  // Response
+  private def matchResponseStatus(expected: InteractionResponse, actual: InteractionResponse): MatchResponseStage[MatchStatusOutcome] =
+    Free.liftF(MatchResponseStatus(expected, actual))
```

Comment:
These functions represent the constructors for the actions you can take and, along with the type declaration above, essentially lift our ADT action into a `Free` context.